### ISSUE TITLE
add StructSR into DiffBir and SeeSr leading to bad performance

### ISSUE THREAD

**Issue by wzh960045** [LYCEXE/StructSR#4]
Sorry ,Do i make any mistake? i cannot get good performance when i  insert this work  into DiffBir and SeeSr
there is my code in DiffBIR sampler:

    def p_sample(
        self,
        model: ControlLDM,total_steps,lamda,SSIM_list,residual_list,cimg,
        x: torch.Tensor,
        model_t: torch.Tensor,
        t: torch.Tensor,
        cond: Dict[str, torch.Tensor],
        uncond: Optional[Dict[str, torch.Tensor]],
        cfg_scale: float,
    ) -> torch.Tensor:
        if lamda==0:
        # predict x_0
            model_output = self.apply_model(model, x, model_t, cond, uncond, cfg_scale)
            if self.parameterization == "eps":
                pred_x0 = self._predict_xstart_from_eps(x, t, model_output)
            else:
                pred_x0 = self._predict_xstart_from_v(x, t, model_output)
            # calculate mean and variance of next state
            mean, variance = self.q_posterior_mean_variance(pred_x0, x, t)
            # sample next state
            noise = torch.randn_like(x)
            nonzero_mask = (t != 0).float().view(-1, *([1] * (len(x.shape) - 1)))
            x_prev = mean + nonzero_mask * torch.sqrt(variance) * noise
            return x_prev
        else:
            lamda= int(total_steps * lamda)
            model_output = self.apply_model(model, x, model_t, cond, uncond, cfg_scale)
            # Structure Condition Embedding (SCE)
            if len(SSIM_list) == lamda:
                struct_cond_2 = residual_list[0]
                cond2=cond
                cond2["c_img"]=struct_cond_2
                uncond2=uncond
                uncond2["c_img"]=struct_cond_2
                model_out_2 = self.apply_model(model,x, model_t, cond2, uncond2, cfg_scale)
                c1 = max(SSIM_list)
                model_output = (1 - c1) * model_output + c1 * model_out_2
            if self.parameterization == "eps":
                pred_x0 = self._predict_xstart_from_eps(x, t, model_output)
            else:
                pred_x0 = self._predict_xstart_from_v(x, t, model_output)
                # Structure-Aware Screening (SAS)
            if len(SSIM_list) < lamda:
                #decode the reconstructed image
                # _, enc_fea_lq = vq_model.encode(im_lq_bs)
                x_samples = model.vae_decode(pred_x0)
                im_sr = torch.clamp((x_samples + 1.0) / 2.0, min=0.0, max=1.0)
                im_sr = im_sr.cpu().numpy().transpose(0, 2, 3, 1) * 255

                # LR image has resized before
                cond_img=model.vae_decode(cond["c_img"])
                im_lr = torch.clamp((cond_img + 1.0) / 2.0, min=0.0, max=1.0)
                im_lr = im_lr.cpu().numpy().transpose(0, 2, 3, 1) * 255

                # SSIM calculation
                SSIM = structural_similarity(im_sr[0], im_lr[0], data_range=255, channel_axis=2)
                print(SSIM)

                # Add to Buffer
                SSIM_list.append(SSIM)
                if max(SSIM_list) == SSIM:

                    #update Z_SE
                    if len(residual_list) == 0:
                        residual_list.append(pred_x0)
                    else:
                        residual_list[0] = pred_x0
            else:
                print(t[0],total_steps,lamda,max(SSIM_list))
                # Image Details Embedding (IDE)
                c2 = t[0]/ (total_steps - lamda)
                c2 = c2 * max(SSIM_list)
                x_reference_1 = residual_list[0]
                pred_x0 = c2 * x_reference_1 + (1 - c2) * pred_x0
            # calculate mean and variance of next state
            mean, variance = self.q_posterior_mean_variance(pred_x0, x, t)
            # sample next state
            noise = torch.randn_like(x)
            nonzero_mask = (t != 0).float().view(-1, *([1] * (len(x.shape) - 1)))
            x_prev = mean + nonzero_mask * torch.sqrt(variance) * noise
            return x_prev
        

    @torch.no_grad()
    def sample(
        self,
        model: ControlLDM,cimg,
        device: str,
        steps: int,
        x_size: Tuple[int],
        cond: Dict[str, torch.Tensor],
        uncond: Dict[str, torch.Tensor],
        cfg_scale: float,
        tiled: bool = False,
        tile_size: int = -1,
        tile_stride: int = -1,
        x_T: torch.Tensor | None = None,
        progress: bool = True,
    ) -> torch.Tensor:
        self.make_schedule(steps)
        self.to(device)
        if tiled:
            forward = model.forward
            model.forward = make_tiled_fn(
                lambda x_tile, t, cond, hi, hi_end, wi, wi_end: (
                    forward(
                        x_tile,
                        t,
                        {
                            "c_txt": cond["c_txt"],
                            "c_img": cond["c_img"][..., hi:hi_end, wi:wi_end],
                        },
                    )
                ),
                tile_size,
                tile_stride,
            )
        if x_T is None:
            x_T = torch.randn(x_size, device=device, dtype=torch.float32)

        x = x_T
        timesteps = np.flip(self.timesteps)
        total_steps = len(self.timesteps)
        iterator = tqdm(timesteps, total=total_steps, disable=not progress)
        bs = x_size[0]
        # SSIM Buffer
        SSIM_list = []
        lamda=0.3
        residual_list = []
        for i, step in enumerate(iterator):
            model_t = torch.full((bs,), step, device=device, dtype=torch.long)
            t = torch.full((bs,), total_steps - i - 1, device=device, dtype=torch.long)
            cur_cfg_scale = self.get_cfg_scale(cfg_scale, step)
            x = self.p_sample(
                model,total_steps,lamda,SSIM_list,residual_list,cimg,  
                x,
                model_t,
                t,
                cond,
                uncond,
                cur_cfg_scale,
            )

        if tiled:
            model.forward = forward
        return x

**Comment by wzh960045** [comment]
and my code in SeeSR pipline and schedular



    # Structure Condition Embedding (SCE)
                if len(SSIM_list) == lamda:
                    print("sce")
                    struct_cond_2 = residual_list[0]
                    struct_cond_2=self.vae.decode(struct_cond_2 / self.vae.config.scaling_factor, return_dict=False)[0]
                    # model_out_2 = self.denoise(args,struct_cond_2,latents,prompt_embeds,ram_encoder_hidden_states,timesteps,num_inference_steps,start_steps,latent_tiled_size, 
                    #                         latent_tiled_overlap,do_classifier_free_guidance,conditioning_scale,guess_mode,batch_size,guidance_scale,cross_attention_kwargs)
                    down_block_res_samples1, mid_block_res_sample1 = [None]*10, None
                    down_block_res_samples1, mid_block_res_sample1 = self.controlnet(
                        controlnet_latent_model_input,
                        t,
                        encoder_hidden_states=controlnet_prompt_embeds,
                        controlnet_cond=struct_cond_2,
                        conditioning_scale=conditioning_scale,
                        guess_mode=guess_mode,
                        return_dict=False,
                        image_encoder_hidden_states = ram_encoder_hidden_states,
                    )


                    if guess_mode and do_classifier_free_guidance:
                        # Infered ControlNet only for the conditional batch.
                        # To apply the output of ControlNet to both the unconditional and conditional batches,
                        # add 0 to the unconditional batch to keep it unchanged.
                        down_block_res_samples1 = [torch.cat([torch.zeros_like(d), d]) for d in down_block_res_samples1]
                        mid_block_res_sample1 = torch.cat([torch.zeros_like(mid_block_res_sample1), mid_block_res_sample1])

                    # predict the noise residual
                    noise_pred1 = self.unet(
                        latent_model_input,
                        t,
                        encoder_hidden_states=prompt_embeds,
                        cross_attention_kwargs=cross_attention_kwargs,
                        down_block_additional_residuals=down_block_res_samples1,
                        mid_block_additional_residual=mid_block_res_sample1,
                        return_dict=False,
                        image_encoder_hidden_states = ram_encoder_hidden_states,
                    )[0]
                    if do_classifier_free_guidance:
                        noise_pred_uncond1, noise_pred_text1 = noise_pred1.chunk(2)
                        noise_pred1 = noise_pred_uncond1 + guidance_scale * (noise_pred_text1 - noise_pred_uncond1)
                    c1 = max(SSIM_list)
                    noise_pred = (1 - c1) * noise_pred + c1 * noise_pred1

                # compute the previous noisy sample x_t -> x_t-1
                latents = self.scheduler.step(noise_pred, self.vae,image,num_inference_steps,lamda,SSIM_list,residual_list,num_inference_steps-i-1,t, latents, **extra_step_kwargs, return_dict=False)[0]

def step(
        self,
        model_output: torch.FloatTensor,vae,cimg,total_steps,lamda,SSIM_list,residual_list,tnum,
        timestep: int,
        sample: torch.FloatTensor,
        generator=None,
        return_dict: bool = True,
    ) -> Union[DDPMSchedulerOutput, Tuple]:
        """
        Predict the sample from the previous timestep by reversing the SDE. This function propagates the diffusion
        process from the learned model outputs (most often the predicted noise).

        Args:
            model_output (`torch.FloatTensor`):
                The direct output from learned diffusion model.
            timestep (`float`):
                The current discrete timestep in the diffusion chain.
            sample (`torch.FloatTensor`):
                A current instance of a sample created by the diffusion process.
            generator (`torch.Generator`, *optional*):
                A random number generator.
            return_dict (`bool`, *optional*, defaults to `True`):
                Whether or not to return a [`~schedulers.scheduling_ddpm.DDPMSchedulerOutput`] or `tuple`.

        Returns:
            [`~schedulers.scheduling_ddpm.DDPMSchedulerOutput`] or `tuple`:
                If return_dict is `True`, [`~schedulers.scheduling_ddpm.DDPMSchedulerOutput`] is returned, otherwise a
                tuple is returned where the first element is the sample tensor.

        """
        t = timestep

        prev_t = self.previous_timestep(t)

        if model_output.shape[1] == sample.shape[1] * 2 and self.variance_type in ["learned", "learned_range"]:
            model_output, predicted_variance = torch.split(model_output, sample.shape[1], dim=1)
        else:
            predicted_variance = None

        # 1. compute alphas, betas
        alpha_prod_t = self.alphas_cumprod[t]
        alpha_prod_t_prev = self.alphas_cumprod[prev_t] if prev_t >= 0 else self.one
        beta_prod_t = 1 - alpha_prod_t
        beta_prod_t_prev = 1 - alpha_prod_t_prev
        current_alpha_t = alpha_prod_t / alpha_prod_t_prev
        current_beta_t = 1 - current_alpha_t

        # 2. compute predicted original sample from predicted noise also called
        # "predicted x_0" of formula (15) from https://arxiv.org/pdf/2006.11239.pdf
        if self.config.prediction_type == "epsilon":
            pred_original_sample = (sample - beta_prod_t ** (0.5) * model_output) / alpha_prod_t ** (0.5)
        elif self.config.prediction_type == "sample":
            pred_original_sample = model_output
        elif self.config.prediction_type == "v_prediction":
            pred_original_sample = (alpha_prod_t**0.5) * sample - (beta_prod_t**0.5) * model_output
        else:
            raise ValueError(
                f"prediction_type given as {self.config.prediction_type} must be one of `epsilon`, `sample` or"
                " `v_prediction`  for the DDPMScheduler."
            )
            # Structure-Aware Screening (SAS)
        if len(SSIM_list) < lamda:
            #decode the reconstructed image
            # _, enc_fea_lq = vq_model.encode(im_lq_bs)
            x_samples =vae.decode(pred_original_sample / vae.config.scaling_factor, return_dict=False)[0]
            im_sr = torch.clamp((x_samples + 1.0) / 2.0, min=0.0, max=1.0)
            im_sr = im_sr.cpu().numpy().transpose(0, 2, 3, 1) * 255

            # LR image has resized before
            im_lr = torch.clamp((cimg + 1.0) / 2.0, min=0.0, max=1.0)
            im_lr = im_lr.cpu().numpy().transpose(0, 2, 3, 1) * 255

            # SSIM calculation
            SSIM = structural_similarity(im_sr[0], im_lr[0], data_range=255, channel_axis=2)
            print(SSIM)

            # Add to Buffer
            SSIM_list.append(SSIM)
            if max(SSIM_list) == SSIM:

                #update Z_SE
                if len(residual_list) == 0:
                    residual_list.append(pred_original_sample)
                else:
                    residual_list[0] = pred_original_sample
        else:
            print(t,tnum,total_steps,lamda,max(SSIM_list))
            # Image Details Embedding (IDE)
            c2 = tnum/ (total_steps - lamda)
            c2 = c2 * max(SSIM_list)
            x_reference_1 = residual_list[0]
            pred_original_sample  = c2 * x_reference_1 + (1 - c2) * pred_original_sample

        # 3. Clip or threshold "predicted x_0"
        if self.config.thresholding:
            pred_original_sample = self._threshold_sample(pred_original_sample)
        elif self.config.clip_sample:
            pred_original_sample = pred_original_sample.clamp(
                -self.config.clip_sample_range, self.config.clip_sample_range
            )

        # 4. Compute coefficients for pred_original_sample x_0 and current sample x_t
        # See formula (7) from https://arxiv.org/pdf/2006.11239.pdf
        pred_original_sample_coeff = (alpha_prod_t_prev ** (0.5) * current_beta_t) / beta_prod_t
        current_sample_coeff = current_alpha_t ** (0.5) * beta_prod_t_prev / beta_prod_t

        # 5. Compute predicted previous sample µ_t
        # See formula (7) from https://arxiv.org/pdf/2006.11239.pdf
        pred_prev_sample = pred_original_sample_coeff * pred_original_sample + current_sample_coeff * sample

        # 6. Add noise
        variance = 0
        if t > 0:
            device = model_output.device
            variance_noise = randn_tensor(
                model_output.shape, generator=generator, device=device, dtype=model_output.dtype
            )
            if self.variance_type == "fixed_small_log":
                variance = self._get_variance(t, predicted_variance=predicted_variance) * variance_noise
            elif self.variance_type == "learned_range":
                variance = self._get_variance(t, predicted_variance=predicted_variance)
                variance = torch.exp(0.5 * variance) * variance_noise
            else:
                variance = (self._get_variance(t, predicted_variance=predicted_variance) ** 0.5) * variance_noise

        pred_prev_sample = pred_prev_sample + variance

        if not return_dict:
            return (pred_prev_sample,)

        return DDPMSchedulerOutput(prev_sample=pred_prev_sample, pred_original_sample=pred_original_sample)

**Comment by wzh960045** [comment]
StructSR code from this github, i am so sorry ,please tell me how i can do
the image become less detail and low IQA
